### ISSUE TITLE
Submit Feature Report with MS4 Work Orders

### ISSUE THREAD

**Issue by cstevenson1** [leeensminger/DelDOT-NPDES-Viewer#417]
@KCI-Ablowers Similar to how BMP Reports are copied to a designated folder for Supreet to consume and attach to the BMP work orders in MAXIMO, DelDOT is interested in attaching the feature report to MS4 work orders. Level of effort may determine how soon it is implemented. I believe @gerrykelly  added a "LINK_TO_INSPECTION_REPORT" field to the submission table for BMPs so we were wondering/hoping a similar approach could be taken for the structure/conveyance reports.

**Comment by cstevenson1** [comment]
@KCI-Ablowers @gerrykelly Matt received approval from Supreet to begin implementing the new process in dev. Similar to BMPs, DelDOT will need to do testing on their end and will want a copy of a dev submission BEFORE we push to production so they can make the necessary changes on their end to consume the data.

**Comment by gerrykelly** [comment]
feature_globalid in 
('E72B817F-0759-445E-88D6-840D011AD184',
'0AC96F87-CD14-4088-8615-FA2C1274A712',
'6047A334-9CBD-4AA3-A2F7-E300C509486C',
'AD2E47E1-3AF6-4C02-B89A-0A29DDB53596')

**Comment by gerrykelly** [comment]
@cstevenson1 @KCI-Ablowers 
(Just the same content as earlier today email)

The new geoprocessing task for bulk export of structure and conveyance (with workorder) reports is deployed at:
https://deldot-agsdev/arcgis/rest/services/DelDotReports/multiReportStructConvey/GPServer/multiReportStructConvey/submitJob
•	I’ll add it to prod ASAP (it doesn’t conflict with anything, so no need to wait on testing in dev)

The process creates the reports and copies them to a network folder, the default being:
[\\sparkssvm\DelDotPics\DelDOTstructure_conveyanceWOreportsForWebservice\Production](file:///DelDotPics/DelDOTstructure_conveyanceWOreportsForWebservice/Production)  (i.e. K:\DelDotPics\DelDOTstructure_conveyanceWOreportsForWebservice\Production)
•	“DelDotPics” as a parent folder is not the best-named, but I don’t think we should put more of these specialized DelDOT folders at the root level on the K-drive
•	There is a … DelDOTstructure_conveyanceWOreportsForWebservice\Development folder also, which we might have use for later, but for now, even though the process is using the dev database, stick with the default production location

I was not sure how you wanted to go about the datasource and record selection for reports to be generated, so what I did is setup the process to query the following fields -- KCIWONO,ASSET_TYPE,ASSET_NUM,FEATURE_GLOBALID – from the selected source.  That means WORK_ORDERS_MS4_Complete_TO_SEND or WORK_ORDERS_MS4_Complete for now, but I can create a custom view if you have some logic that works better.
•	The submit-job page is like below:
![image](https://user-images.githubusercontent.com/8259136/159978422-46a3c06f-3bc4-4b13-ac06-a927a0eed4d3.png)
o	With “where criteria:” blank, you simply get all record from the selected source
o	The where criteria can be used just like you would in the ArcMap select by attributes criteria, like
![image](https://user-images.githubusercontent.com/8259136/159978474-7acde488-f89b-4b92-931f-ca285b351114.png)
o	The task only returns the most basic messaging, so from the above you would see:
o	esriJobMessageTypeInformative: 4 to be generated; 2 conveyances & 2 structures
o	esriJobMessageTypeInformative: CONVEYANCE report 1 of 2 ASSET_NUM-0112317395 (Roadway Culvert): 48188791-cc6a-4b34-b5f5-06ccf486481a.pdf
o	esriJobMessageTypeInformative: CONVEYANCE report 2 of 2 ASSET_NUM-0111107636 (Pipe): b396e56b-99e9-4a59-a1fa-02630bb32466.pdf
o	esriJobMessageTypeInformative: STRUCTURE report 1 of 2 ASSET_NUM-0112056154 (Inlet): 63bfb1ec-5538-4eb0-b6be-0943ad7f6a46.pdf
o	esriJobMessageTypeInformative: STRUCTURE report 2 of 2 ASSET_NUM-0109045514 (Junction Box): 47994c47-ce3d-48fe-b6a2-08c3028f4985.pdf
o	The optional parameters are placeholders if needed for future revisions

Please test.

**Comment by gerrykelly** [comment]
@cstevenson1 @KCI-Ablowers
(Just the same content as earlier today email)

For the new webservice download option for structure and conveyance (with workorder) reports, I’ve:
•	Added a field -- [LINK_TO_REPORT] -- to WORK_ORDERS_MS4_Complete_TO_SEND & WORK_ORDERS_MS4
o	Values like: reportfilename=EE26912B-62B3-41CF-BFB2-09DAD4CCEECB.pdf&env=Dev&f=pjson
•	Updated the https://newarkgis.kci.com/reportstreamer/stream.asmx/GetFile routine to return the base64 version of the report file, just as is done for the BMP reports
o	The request follows the same logic, just make the URL query string whatever is in the link field:
https://newarkgis.kci.com/reportstreamer/stream.asmx/GetFile?reportfilename=B396E56B-99E9-4A59-A1FA-02630BB32466.pdf&env=Dev&f=pjson

If you get a (dev data based) table from  WORK_ORDERS_MS4_Complete* to Supreet, I think it would be a quick and successful test for him to just try whatever he does for the BMP reports, using the new link values.

I don’t think anyone but me needs Python test client code I provided when we were working to get the BMP report download to work, but here it is below in case it helps:

import requests, base64, xml.etree.ElementTree as ET,   urllib; from urlparse import parse_qs, urlparse
url = 'https://newarkgis.kci.com/reportstreamer/stream.asmx/GetFile?reportfilename=B396E56B-99E9-4A59-A1FA-02630BB32466.pdf&env=Dev&f=pjson'
filename= r"c:/temp/" + parse_qs(urlparse(url).query)["reportfilename"][0]  
response = requests.get(url, verify=False)
base64text = ET.fromstring(response.content).text 
base64bytes = base64.b64decode(base64text)
open(filename, 'wb').write(base64bytes)


(note that I slightly modified it since last time ignore authenticating the server’s SSL certificate.  If Supreet gets a certificate-related error, he might need to do the same on his end, at least temporarily. )

**Comment by gerrykelly** [comment]
@cstevenson1 @KCI-Ablowers 

I published the service to prod:
https://newarkgis.kci.com/arcgis/rest/services/DelDotReports/multiReportStructConvey/GPServer/multiReportStructConvey/submitJob

I'm testing it by having "Save reports to folder as";  \\sparkssvm\DelDotPics\DelDOTstructure_conveyanceWOreportsForWebservice\Production_Testing
- That folder being there just for our testing and going forward you would use the \Production folder for the process,
- It looks like it is going to successfully create 395 reports for the 395 records in prod WORK_ORDERS_MS4_Complete_TO_SEND

I'm removing my assignment now.  If anything comes up with the GP service just assigned back to me; if anything comes up with Supreet and the web service, make that a new issue and link to this.

**Comment by cstevenson1** [comment]
@gerrykelly will do. I imagine it will be another week; it's been passed off to the client for testing to allow them to make necessary updates on their end to consume the reports.

**Comment by KCI-Ablowers** [comment]
Closing for now and will reopen if any issues arise from Supreet/DelDOT
